### PR TITLE
add the logic for report generation

### PR DESCRIPTION
**Pull request by seyedparsa**
The logic of report generation has been implemented in the report models file. There are also some minor changes in the core models file.

### CODE REVIEW

File: roadservice/reporting/models.py [hunk 1]
```diff
@@ -1,178 +1,111 @@
...
+        if cls.__instance is None:
+            cls()
+        return cls.__instance
+
+    def __init__(self):
+        if ReportGenerator.__instance is not None:
+            raise SingletonInitError()
+        ReportGenerator.__instance = self
+
+    @staticmethod
```
Comment (kgolezardi) [inline roadservice/reporting/models.py]: If it's static, why did we need an instance of `ReportGenerator`?

File: roadservice/reporting/models.py [hunk 2]
```diff
@@ -1,178 +1,111 @@
...
+        SUBREGIONS_REPORT = 'S'
+
+
+class IntervalReport(Report):
+    def __init__(self, region, start_time, end_time):
+        self.region = region
+        self.start_time = start_time
+        self.end_time = end_time
+        self.interval_delta = timezone.timedelta(weeks=1)
+        self.bins = []
```
Comment (kgolezardi) [inline roadservice/reporting/models.py]: `bin_starts`

File: roadservice/reporting/models.py [hunk 3]
```diff
@@ -1,178 +1,111 @@
...
+            bin_start = time
+            bin_end = time + report.interval_delta
+            time += report.interval_delta
+            report.bins.append(bin_start)
+            report.team_counts.append(region.get_teams().filter(
+                Q(created_at__lt=bin_end) & (Q(deleted_at__gt=bin_start) | Q(deleted_at__isnull=True))).count())
+            issues = region.get_issues().filter(created_at__gte=bin_start, created_at__lt=bin_end)
+            report.issue_counts.append(issues.count())
+            report.failed_issue_counts.append(issues.filter(state=Issue.State.FAILED).count())
+            report.done_issue_counts.append(issues.filter(state=Issue.State.DONE).count())
```
Comment (kgolezardi) [inline roadservice/reporting/models.py]: Check if done or scored.

File: roadservice/reporting/models.py [hunk 4]
```diff
@@ -1,178 +1,111 @@
...
+            missions = region.get_missions().filter(issue__created_at__gte=bin_start, issue__created_at__lt=bin_end)
+            report.score_averages.append(missions.aggregate(Avg('score'))['score__avg'])
```
Comment (kgolezardi) [inline roadservice/reporting/models.py:80]: How does it behave when some rows include `null`?

File: roadservice/reporting/models.py [hunk 5]
```diff
@@ -1,178 +1,111 @@
...
+            issues = subregioin.get_issues()
+            mission_count = missions.count()
+            issue_count = issues.count()
+            mission_score_avg = missions.aggregate(Avg('score'))['score__avg']  # TODO
+            successful_issues_count = issues.filter(Q(state=Issue.State.DONE) | Q(state=Issue.State.SCORED)).count()
+            failed_issues_count = issues.filter(state=Issue.State.FAILED).count()
+            finished_issues_count = successful_issues_count + failed_issues_count
+            mission_success_rate = None if (finished_issues_count == 0) else \
+                successful_issues_count / finished_issues_count
+            report.subregions_info.append((subregioin, mission_count, issue_count,
```
Comment (kgolezardi) [inline roadservice/reporting/models.py]: Typo in name

File: roadservice/reporting/models.py [hunk 6]
```diff
@@ -1,178 +1,111 @@
...
+        self.speciality_dist = defaultdict(int)
+        self.machinery_type_dist = defaultdict(int)
+        self.score_dist = defaultdict(int)
+        self.Type = Report.Type.DISTRIBUTION_REPORT
+
+
+class SubregionsReport(Report):
+    def __init__(self, region):
+        self.region = region
+        self.subregions_info = []
```
Comment (kgolezardi) [inline roadservice/reporting/models.py]: Would be nice to add a comment to specify the structure of this list.

File: roadservice/reporting/models.py (hunk 5 above)
Comment (kgolezardi) [reply roadservice/reporting/models.py]: Also missing team count.

**Review by kgolezardi — CHANGES_REQUESTED** [review]

File: roadservice/reporting/models.py (hunk 4 above)
Comment (seyedparsa) [reply roadservice/reporting/models.py:80]: That's OK! It considers the rows which are not null and returns None if all of them are null.

**Review by kgolezardi — APPROVED** [review]

**Pull request merged**